fblock.go

```diff
@@ -0,0 +1,455 @@
+	// there will be 10 of them.  This may change in the future.
+	endOfPeriod [10]int
+}
+
+type FBlockHeader struct {
+	BodyMR          *Bytes32 `json:"bodymr"`
+	PrevKeyMR       *Bytes32 `json:"prevkeymr"`
+	PrevLedgerKeyMR *Bytes32 `json:"prevkeymr"`
+
+	ExchangeRate uint64 `json:"exchrate"`
+	Height       uint32 `json:"dbheight"`
+
+	// ExpansionBytes is the expansion space in the fblock. If we do not understand the
+	// expansion, we just store the raw bytes
+	ExpansionSize  int64 `json:"expansionsize"` // Stored as a varint
+	ExpansionBytes Bytes `json:"expansiondata"`
+
+	// Number of txs in the block
+	TransactionCount uint32 `json:"txcount"`
+	// Number of bytes contained in the body
+	BodySize uint32 `json:"bodysize"`
+}
+
+// IsPopulated returns true if fb has already been successfully populated by a
+// call to Get. IsPopulated returns false if fb.Transactions is nil.
+func (fb FBlock) IsPopulated() bool {
+	return len(fb.Transactions) > 0 && // FBlocks always contain at least a coinbase
+		fb.Header.BodyMR != nil &&
+		fb.Header.PrevKeyMR != nil && // TODO: Why not check the computed keymr?
+		fb.Header.PrevLedgerKeyMR != nil
+}
+
+// Get queries factomd for the Factoid Block at fb.Header.Height or fb.KeyMR.
+// After a successful call, the Transactions will all be populated.
+func (fb *FBlock) Get(c *Client) (err error) {
+	if fb.IsPopulated() {
+		return nil
+	}
+
+	if fb.KeyMR != nil {
+		orig := fb.KeyMR
+		defer func() { // Ensure we got the block we asked for
+			if err != nil {
+				return
+			}
+
+			if fb.KeyMR != nil && *orig != *fb.KeyMR {
+				err = fmt.Errorf("invalid key merkle root")
+				return
+			}
+		}()
+
+		params := struct {
+			Hash *Bytes32 `json:"hash"`
+		}{Hash: fb.KeyMR}
+		var result struct {
+			Data Bytes `json:"data"`
+		}
+		if err := c.FactomdRequest("raw-data", params, &result); err != nil {
+			return err
+		}
+		return fb.UnmarshalBinary(result.Data)
+	}
+
+	params := struct {
+		Height uint32 `json:"height"`
+	}{fb.Header.Height}
+	result := struct {
+		// We will ignore all the other fields, and just unmarshal from the raw.
+		RawData Bytes `json:"rawdata"`
+	}{}
+	if err := c.FactomdRequest("fblock-by-height", params, &result); err != nil {
+		return err
+	}
+
+	return fb.UnmarshalBinary(result.RawData)
+}
+
+const (
+	FBlockMinHeaderLen = 32 + // Factoid ChainID
+		32 + // BodyMR
+		32 + // PrevKeyMR
+		32 + // PrevLedgerKeyMR
+		8 + // EC Exchange Rate
+		4 + // DB Height
+		0 + // Header Expansion size (varint)
+		0 + // Header Expansion Area (Min 0)
+		4 + // Transaction Count
+		4 // Body Size
+
+	FBlockMinTotalLen = FBlockMinHeaderLen
+)
+
+// UnmarshalBinary unmarshals raw directory block data.
+//
+// Header
+// [Factoid Block ChainID (Bytes32{31:0x0a})] +
+// [BodyMR (Bytes32)] +
+// [PrevKeyMR (Bytes32)] +
+// [PrevLedgerKeyMR (Bytes32)] +
+// [Exchange Rate (8 bytes)] +
+// [DB Height (4 bytes)] +
+// [Header Expansion size (Bytes)] +
```

Comment:
I do not have the answer to that offhand. I'd have to bring that question up with Brian and Paul.

Reply:
So just something I learned recently. An expansion thing we might add is a balancehash for leaders to check their balances against each other. This would be an optional thing for nodes to check against, and mainly for the authority set.

So instead of making a hard fork, we can use the expansion space for a feature like this, without impacting any running nodes.